File: trunk/tools/setenv/setenv.pl
#! /usr/bin/env perl
# -*-Mode: perl;-*-

# -*-Mode: perl;-*-
# make sure we execute perl
#eval '(exit $?0)' && eval 'exec perl -S $0 ${1+"$@"}'
#& eval 'exec perl -S $0 $argv:q'
#if 0;

# $Header: /m_home/m_utkej/Argonne/cvs2svn/cvs/OpenAD/tools/setenv/setenv.pl,v 1.2 2004-05-21 13:54:10 eraxxon Exp $
## * BeginCopyright *********************************************************
## 
## 
## *********************************************************** EndCopyright *

#############################################################################
##
## $Source: /m_home/m_utkej/Argonne/cvs2svn/cvs/OpenAD/tools/setenv/setenv.pl,v $ 
##
##   Nathan Tallent.
##
#############################################################################

use strict;
use warnings;

use FindBin qw($RealBin);
use Cwd qw(abs_path);
use IO::File;
use Getopt::Long;

use lib "$RealBin/../libperl";
use setenv qw(genSetEnvVar genSetVar genAppendEnvVar 
	      genUnSetEnvVar genUnSetVar genPrintEnvVar genPrintVar
	      genSetAlias genUnSetAlias genIf
	      is_sh is_csh);

use lib "$RealBin/../..";
use openad_config;

my $OpenADRoot = abs_path("$RealBin/../..");

STDOUT->autoflush(1);

#############################################################################

my $the_program = "setenv.pl";
my $the_usage = 
"usage: ${the_program} [options] <shell>

  Required arguments:
    -s, --shell=<shell-type> : sh | csh

  General options:
    -h, --help : print help
\n";

my @the_options = ('shell|s=s',
		   'help|h',
		   );

#############################################################################
## main/driver
#############################################################################

my $the_homedir = $ENV{'HOME'};

my $opt_shell = undef;
my $opt_category = undef;


# ----------------------------------------------------------
# Parse the command line
# ----------------------------------------------------------

parseOptions($0);

# ----------------------------------------------------------
# Run
# ----------------------------------------------------------

GenEnvSettings($opt_shell);

#############################################################################
## GenEnvSettings
#############################################################################

my %platformToOpen64TargTable;
my %varidx;
my @RootEnvVars;
my @OtherEnvVars;
my @Aliases;

BEGIN {

  %platformToOpen64TargTable =
      (
       'alpha-OSF1'  => 'targ_alpha_tru64', 
       'i686-Linux'  => 'targ_ia32_ia64_linux',
       'i686-Cygwin' => 'targ_ia32_ia64_linux',
       'ia64-Linux'  => 'targ_ia64_ia64_linux',
       'mips-IRIX64' => 'targ_mips_irix',
       'sparc-SunOS' => 'targ_sparc_solaris',
       );
  
  %varidx = ( 'var' => 0, 'val' => 1, );
  
  @RootEnvVars =
      (
       ['OPEN64ROOT',       '${OPEN64_BASE}/osprey1.0/${o64targ}'],
       ['OPENANALYSISROOT', '${OPENANALYSIS_BASE}/${PLATFORM}'],
       ['XERCESCROOT',      '${XERCESC_BASE}/${PLATFORM}'],
       ['XAIFBOOSTERROOT',  '${XAIFBOOSTER_BASE}'],
       ['BOOSTROOT',        '${BOOST_BASE}'],
       ['ANGELROOT',        '${ANGEL_BASE}'],
       ['XAIFSCHEMAROOT',   '${XAIFSCHEMA_BASE}'],
       [ undef,             undef ],
       );
  
  @OtherEnvVars =
      (
       ['OPENADFORTTK_OPEN64BASE', '${OPEN64_BASE}/osprey1.0'],
       ['OPENADFORTTK_OPEN64',     '${OPEN64ROOT}'],
       ['OPENADFORTTK_OA',         '${OPENANALYSISROOT}'],
       ['OPENADFORTTK_XERCESC',    '${XERCESCROOT}'],
       ['OPENADFORTTK_BOOST',      '${BOOSTROOT}'],
       [ undef,                    undef],
       [ 'XAIF_DIR',               '${XAIFBOOSTERROOT}' ],
       [ 'BOOST_DIR',              '${BOOSTROOT}' ],
       [ undef,                    undef],
       );
  
  @Aliases =
      (
       ['mfef90',    '${OPEN64ROOT}/crayf90/sgi/mfef90' ],
       ['whirl2f',   '${OPEN64ROOT}/whirl2f/whirl2f' ],
       ['whirl2f90', '${OPEN64ROOT}/whirl2f/whirl2f90'],
       ['ir_b2a',    '${OPEN64ROOT}/ir_tools/ir_b2a' ],
       ['ir_size',   '${OPEN64ROOT}/ir_tools/ir_size'],
       );
}

# GenEnvSettings: Generate code to setup CVS repository variables
# returns: 
# effect: 
# assumes:
sub GenEnvSettings
{
  my($shell) = @_;
  
  # --------------------------------------------------------
  # Generate canonical platform
  # --------------------------------------------------------
  my $platform = 'i686-Linux'; # `cd ${OpenADRoot}/config; ./hpcplatform`;
  print STDOUT genSetVar('PLATFORM', $platform, $shell);

  my $o64targ = $platformToOpen64TargTable{$platform};
  if (!defined($o64targ)) {
    die "Programming error: Unknown platform!\n";
  }
  print STDOUT genSetVar('o64targ', $o64targ, $shell);

  # --------------------------------------------------------
  # Generate BASE vars for sub repositories
  # --------------------------------------------------------
  print STDOUT "\n";

  my $config = openad_config->new();
  my $OpenADRepos = $config->getRepos();

  for my $repo (@{$OpenADRepos}) {
    my $repoPath = $repo->{path} . '/' . $repo->{name};
    if (-d $repoPath) {
      my $var = $repo->{var};
      my $val = $repoPath;
      print STDOUT genSetEnvVar($var, $val, $shell);
      #print STDOUT genPrintEnvVar($var, $shell);
    }
  }

  # --------------------------------------------------------
  # Generate environment vars for sub repositories
  # --------------------------------------------------------
  print STDOUT "\n";

  my @EnvVars = (@RootEnvVars, @OtherEnvVars);
  for my $pair (@EnvVars) {
    my $var = $pair->[$varidx{var}];
    my $val = $pair->[$varidx{val}];
    if (defined($var)) {
      print STDOUT genSetEnvVar($var, $val, $shell);
    } else {
      print STDOUT "\n";
    }
  }

  # --------------------------------------------------------
  # Generate other environment stuff
  # --------------------------------------------------------
  
  if ($platform eq 'i686-Cygwin') {
    my $path = '${XERCESCROOT}/bin:${XERCESCROOT}/lib:${OPEN64ROOT}/be:${OPEN64ROOT}/whirl2f:${PATH}';
    print STDOUT genAppendEnvVar('PATH', $path, $shell);
  } else {
    my $ldlib = '${XERCESCROOT}/lib:${OPEN64ROOT}/whirl2f';
    print STDOUT genAppendEnvVar('LD_LIBRARY_PATH', $ldlib, $shell);
  }    
  
  
  # --------------------------------------------------------
  # Generate aliases
  # --------------------------------------------------------
  print STDOUT "\n";

  for my $pair (@Aliases) {
    my $var = $pair->[$varidx{var}];
    my $val = $pair->[$varidx{val}];
    print STDOUT genSetAlias($var, $val, $shell);
  }
  
  # --------------------------------------------------------

  print STDOUT "\n";
  print STDOUT genUnSetVar('PLATFORM', $shell);
  print STDOUT genUnSetVar('o64targ', $shell);


}

#############################################################################
## parseOptions
#############################################################################

sub parseOptions
{
  my ($command) = @_;
  
  # Get optional arguments
  my %opts = ();
  my $ret = GetOptions(\%opts, @the_options);
  if (!$ret) { 
    printErrorAndExit(); 
  }
  
  # Get optional arguments: help
  if (defined( $opts{'help'} )) {
    printUsageAndExit($command);
  }
  
  # ----------------------------------------------------------
  # Required arguments
  # ----------------------------------------------------------

  # Shell type
  if (defined($opts{'shell'})) {
    $opt_shell = $opts{'shell'};
  } else {
    printErrorAndExit("Shell option missing.\n");
  }
  if (! (is_sh($opt_shell) || is_csh($opt_shell)) ) {
    printErrorAndExit("Invalid shell '$opt_shell'.\n");
  }
  
  ## Make sure called with minimum number of required arguments
  my $numArgs = scalar(@ARGV);
  if ($numArgs != 0) { 
    printErrorAndExit("Invalid number of required arguments!\n");
  }
  
  #print STDOUT "shell:           $opt_shell\n";
}

# printUsageAndExit
sub printUsageAndExit 
{
  my ($command) = @_; # not used now
  print STDOUT ${the_usage};
  exit(-1);
}

# printErrorAndExit
sub printErrorAndExit 
{
  my ($msg) = @_;
  if (defined($msg)) {
    print STDOUT "${msg}";
  }
  print STDOUT "Try `${the_program} --help' for more information.\n";
  exit(-1);
}

#############################################################################

# Local Variables:
# perl-indent-level: 2
# End:

File: trunk/tools/setenv/setenv.py
#!/usr/bin/env python
##########################################################
# This file is part of OpenAD released under the LGPL.   #
# The full COPYRIGHT notice can be found in the top      #
# level directory of the OpenAD distribution             #
##########################################################

###############################################################################
import os
import sys
import warnings
import subprocess
import getopt
# Assumes executing script is in main OpenAD directory
mydir = os.path.dirname(__file__)
OpenADRoot = os.path.abspath(os.path.join(mydir, "../.."))
sys.path.append(OpenADRoot)
import openadConfig
libsetenvdir = os.path.join(OpenADRoot, "tools/libpythontk")
sys.path.append(libsetenvdir)
import libsetenv

class setenv:
  def __init__(self):

###############################################################################
    self.the_program = os.path.basename(sys.argv[0])
    self.the_usage='usage: '+self.the_program+' [options] <shell> \
        Required arguments: \
        -s, --shell=<shell-type> : sh | csh\
    \
      General options:\
        -h, --help : print help\
      \n'

    self.the_options = 'shell:s=s,help:h'
    self.opt_shell = 'undef'
    self.help ="Try "+self.the_program+" --help' for more information.\n"


###############################################################################
# parse the command line
###############################################################################

    self.parseCmdLine(sys.argv) #assigns self.opt_shell, or exits on error
    self.libsetenv_instance = libsetenv.libsetenv(self.opt_shell)

###############################################################################
# GenEnvSettings
###############################################################################

# GenEnvSettings: Generate code to setup CVS repository variables
  def GenEnvSettings(self):
    config=openadConfig.openadConfig()
    OpenADRepos = config.getRepos()
    # BASE environment variables
    print self.libsetenv_instance.genSetEnvVar('OPENAD_BASE',OpenADRoot)
    for key,repoPair in (OpenADRepos.items()):
      if repoPair[0].getVar() is not None:
        print self.libsetenv_instance.genSetEnvVar(repoPair[0].getVar(),os.path.join(repoPair[0].getLocalPath(),repoPair[0].getLocalName()))
    # ROOT environment variables
    print self.libsetenv_instance.genSetEnvVar('OPENADROOT', OpenADRoot)
    for var,val in config.RootEnvVars.items():
      try:
        print self.libsetenv_instance.genSetEnvVar(var,val)
      except NameError, e:
        print e
        pass
    # PATH and LD_LIBRARY_PATH
    print self.libsetenv_instance.genAppendEnvVar('PATH',os.path.join(os.environ['OPENADFORTTK']+'bin'))
    print self.libsetenv_instance.genAppendEnvVar('PATH',os.path.join(os.environ['OPENADROOT'],'bin'))
    if(config.platform=='i686-Cygwin' or config.platform=='x86-Cygwin'):
      path = os.environ['XERCESCROOT']+'/bin:'+os.environ['XERCESCROOT']+'/lib:'+os.environ['OPEN64ROOT']+'/be:'+os.environ['OPEN64ROOT']+'/whirl2f:'+os.environ['PATH']
      print self.libsetenv_instance.genAppendEnvVar('PATH', path)
    else:
      ldlib=os.environ['OPEN64ROOT']+'/whirl2f'
      ldPathName='LD_LIBRARY_PATH'
      if (config.platform=='x86-MacOS'):
        ldPathName='DY'+ldPathName
      print self.libsetenv_instance.genAppendEnvVar(ldPathName, ldlib)
    # aliases
    for var,val in config.Aliases.items():
      try:
        command = self.libsetenv_instance.genSetAlias(var, val)
        print command
      except NameError, e:
        print e
        pass

# parseCmdLine: process args and assign self.opt_shell variable, or print error message
  def parseCmdLine(self, command):
    # Get optional arguments
    opts=[]
    opts,ret = getopt.getopt(command[1:],'s:h',['shell=','help'])
    ## the getopt call  removes options from ARGV and 
    ## there should be nothing left in ret.
    if len(ret) != 0:
      self.printErrorAndExit()
    # Get optional arguments: help
      try:
        self.printUsageAndExit()
      except RuntimeError, e:
        print e
        pass

  # ----------------------------------------------------------
  # Required arguments
  # ----------------------------------------------------------
  # Shell type
    for opt in opts:
      arg, self.opt_shell = opt
      if arg == "--shell":
        break
    if self.opt_shell == 'undef':
      self.printErrorAndExit("Shell option missing\n")
    testShell = libsetenv.libsetenv(self.opt_shell)
    if not (testShell.is_sh() or testShell.is_csh()):
       self.printErrorAndExit("Shell "+self.opt_shell+" is not a sh or csh type shell.\n")



  def printUsageAndExit(self):
    print self.the_usage
    sys.exit()

  def printErrorAndExit(self,msg=help):
    print msg
    sys.exit()


# ----------------------------------------------------------
# Run
# ----------------------------------------------------------

setenv=setenv()
setenv.GenEnvSettings()
